rust/arrow/src/json/reader.rs

```diff
@@ -330,136 +307,237 @@ pub fn infer_json_schema<R: Read>(
+                            field_types.insert(
+                                k.to_string(),
+                                InferredType::Array(Box::new(InferredType::Any)),
+                            );
+                        }
+                    }
+                }
+
+                match field_types.get_mut(k).unwrap() {
+                    InferredType::Array(inner_type) => {
+                        inner_type.merge(ele_type)?;
+                    }
+                    // in case of column contains both scalar type and scalar array type, we
+                    // convert type of this column to scalar array.
+                    field_type @ InferredType::Scalar(_) => {
+                        field_type.merge(ele_type)?;
+                        *field_type = InferredType::Array(Box::new(field_type.clone()));
+                    }
+                    t => {
+                        return Err(ArrowError::JsonError(format!(
+                            "Expected array json type, found: {:?}",
+                            t,
+                        )));
+                    }
+                }
+            }
+            Value::Bool(_) => {
+                set_object_scalar_field_type(field_types, k, DataType::Boolean)?;
+            }
+            Value::Null => {
+                // do nothing, we treat json as nullable by default when
+                // inferring
+            }
+            Value::Number(n) => {
+                if n.is_f64() {
+                    set_object_scalar_field_type(field_types, k, DataType::Float64)?;
+                } else {
+                    // default to i64
+                    set_object_scalar_field_type(field_types, k, DataType::Int64)?;
+                }
+            }
+            Value::String(_) => {
+                set_object_scalar_field_type(field_types, k, DataType::Utf8)?;
+            }
+            Value::Object(inner_map) => {
+                if !field_types.contains_key(k) {
+                    field_types
+                        .insert(k.to_string(), InferredType::Object(HashMap::new()));
+                }
+                match field_types.get_mut(k).unwrap() {
+                    InferredType::Object(inner_field_types) => {
+                        collect_field_types_from_object(inner_field_types, inner_map)?;
+                    }
+                    t => {
+                        return Err(ArrowError::JsonError(format!(
+                            "Expected object json type, found: {:?}",
+                            t,
+                        )));
+                    }
+                }
+            }
+        }
+    }
+
+    Ok(())
+}
+
 /// Infer the fields of a JSON file by reading all items from the JSON Value Iterator.
+///
+/// The following type coercion logic is implemented:
+/// * `Int64` and `Float64` are converted to `Float64`
+/// * Lists and scalars are coerced to a list of a compatible scalar
```

Comment:
This line (appears elsewhere) should no longer apply when following same behaviour as Spark